### PR TITLE
fix(extract): preflight co-importability check with actionable duplicate-declaration diagnostic

### PR DESCRIPTION
**Pull request by jinxinglim**
# Issue

Running `probe-lean extract .` on a project where multiple modules define the same top-level names — e.g. benchmark-style repos shipping `H1/problem.lean` and `H1/solution.lean` as near-identical copies without namespaces — fails at the atomize step (#61). Lake compiles each module independently so the project builds fine, but atomize imports **all** built modules into a single Lean environment, and the kernel rejects the second definition:

```
Analysis failed: Failed to import modules: import H1.solution failed, environment already contains 'Admissible' from H1.problem
```

Worse, the error hint blamed a stale orphan `.olean` and told the user to run `lake clean` — a dead end for this case (it reproduces after `rm -rf .lake/`), since orphan oleans have been filtered by source existence since #52.

Closes #61
Closes #62

# Solution

Per the analysis in #62, co-importability is a **hard project requirement**, not a probe-lean bug: verilib runs probe-lean automatically with no per-project knowledge (so `-m`/`-l` workarounds don't exist on that path), and lakefile-level grouping (`defaultTargets`, `[[lean_lib]]` splits) provably cannot avoid the clash. What *is* a probe-lean bug is the diagnostic. Following #62's two requests:

1. **Documented requirement** — README **Supported Projects** gains a *Co-importable modules* row and **What won't work** gains a bullet with the failure signature; the USAGE.md troubleshooting entry that wrongly suggested `defaultTargets` as a fix is corrected.
2. **Preflight diagnostic** — before the expensive import, `extract` reads each built module's own declarations from its `.olean` header (`Lean.readModuleData`, the approach validated by the reference script in #62) and, on a collision, aborts with the duplicated names, their owning modules, the structural fixes, and a `--module` escape hatch — no more `lake clean` misdirection.

Two refinements beyond the reference script:

- The check replicates the importer's duplicate-tolerance rule (`subsumesInfo`): identical-statement theorem/axiom restatements are **exempt**, so the preflight never rejects a project the importer would accept. (Deliberately lenient on the axiom/axiom Prop check, which needs the full constant map — leniency only under-detects.)
- Known misses (collisions involving dependency modules, module-system split parts, unreadable oleans) are under-detection only: the import then fails as before and lands on a retained fallback hint that now covers both the stale-olean and duplicate-source causes.

# Flow of the implementation

1. **Discovery — `ProbeLean/Environment.lean`**: `getProjectModules` now returns `ProjectModule` records — each module name stapled to the `.olean` path it was discovered from. `partitionBySource` keeps `(name, relPath)` pairs so the pairing is never re-derived after filtering.
2. **Filtering — `ProbeLean/Extract.lean`**: `selectModules` filters whole records, so `--library`/`--module` can never mispair a module with its olean. This is also why `extract . -m H1.solution` works on a conflicting project: the conflicting sibling is removed before the check runs.
3. **Preflight — `ProbeLean/Coimport.lean`** (new): reads each module's declared names + constants from the olean header (cheap mmap, project modules only — dependency oleans are never touched); `findCoimportCollisions` flags names owned by more than one module, `constSubsumes` applies the importer's exemption, `formatCoimportError` renders the diagnostic (sorted, capped at 10 with an "… and N more" tail, internal names hidden unless they are the only evidence, `--module` example chosen to avoid the prefix trap for root/submodule clashes). Unreadable oleans are skipped, warned about, and named in the output so a partial scan never poses as complete.
4. **Gate — `ProbeLean/Atomize.lean`**: `runAnalysisViaLakeEnv` runs the preflight right before `importModules`; collisions → exit 1 with the report. If the import still fails with "already contains" despite a clean preflight, the fallback hint covers the remaining causes (including the genuine stale-olean case, where `lake clean` is still suggested).
5. **Tests — `Tests/Main.lean`**: 59 new/updated assertions — every branch of the subsumption rule, the collision core (dedup, sorting, exemptions, declared-name keying, internal names), the diagnostic content (including truncation honesty and the hidden-only edge case), and path-preservation through filters.

The decision logic is pure (no IO), which is what makes it exhaustively unit-testable; the IO shell only reads oleans and prints.

# How to test

```bash
lake build && lake build tests && .lake/build/bin/tests   # 585 passed, 0 failed
```

Quick colliding fixture (no Mathlib, ~30s):

```bash
mkdir /tmp/colliding && cd /tmp/colliding
cp <probe-lean>/lean-toolchain .
printf 'name = "fixture"\ndefaultTargets = ["Problem", "Solution"]\n\n[[lean_lib]]\nname = "Problem"\n\n[[lean_lib]]\nname = "Solution"\n' > lakefile.toml
printf 'def Admissible (n : Nat) : Prop := n > 0\ntheorem shared_thm : 1 + 1 = 2 := rfl\n' | tee Problem.lean > Solution.lean
probe-lean extract .   # aborts in preflight, exit 1
```

Note `Admissible` is reported but the identically-restated `shared_thm` is correctly exempt.

Real-project check on [zeta-h123](https://github.com/AxiomMath/zeta-h123) (Lean v4.28.0 — needs a matching build):

```bash
probe-lean extract . -o /tmp/zeta.json                    # new diagnostic, exit 1
probe-lean extract . -m H1.solution -o /tmp/h1sol.json    # workaround: 42 atoms, exit 0
```

Sample console output on zeta-h123 (was: `environment already contains 'Admissible'` + the misleading `lake clean` hint):

```
Analysis failed: Co-importability check failed: 24 declaration name(s) are declared by more than one module.

  AddNoCarry — declared in: H2.problem, H2.solution
  Admissible — declared in: H1.problem, H1.solution
  CarryFree — declared in: H1.problem, H1.solution, Lem41.problem, Lem41.solution
  F — declared in: H1.problem, H1.solution, H2.problem, H2.solution
  ...
  … and 14 more duplicated name(s)

probe-lean imports all built modules into a single Lean environment. Lake
compiles each module independently, so the project builds — but Lean forbids
duplicate declarations in one environment, so extraction cannot proceed.

Fix the project structure: give each variant family its own namespace, or
have the dependent module `import` the shared module instead of restating
its definitions.

For a manual run, a non-conflicting subset can be extracted with --module,
e.g.: probe-lean extract . --module H2.problem
(--module also selects submodules of the named module; --library matches module-name roots, not lakefile library names).

See README "Supported Projects" for the co-importability requirement.
```

Regression check: on co-importable projects the change is invisible — extract output on curve25519-dalek-lean-verify (2039 atoms, Lean v4.30.0-rc2) is identical to main's modulo the version and timestamp fields.

### CODE REVIEW

File: docs/USAGE.md [hunk 1]
```diff
@@ -78,6 +78,8 @@ probe-lean extract <PROJECT_PATH> [OPTIONS]
 | `--skip-enrich` | | Skip transitive verification enrichment (no `"transitively-verified"` status) |
 | `--class <NAME>` | | Override the detected project class (e.g. `security-protocol`). Normally auto-detected from a VCVio dependency; this is for manual runs |
 
+Before importing, `extract` runs a **co-importability preflight**: it reads each built module's own declarations from its `.olean` header and aborts with the list of duplicated names and their owning modules if two modules declare the same fully-qualified name (see [Troubleshooting](#co-importability-check-failed)). This catches problem/solution-style layouts before the expensive import instead of failing mid-import with a raw kernel error.
```
Comment (astefano) [inline docs/USAGE.md]:
```suggestion
Before importing, `extract` runs a **co-importability preflight**: it reads each built module's own declarations from its `.olean` header and aborts with the list of duplicated names and their owning modules if two modules declare the same fully-qualified name (see [Troubleshooting](#co-importability-check-failed)).
```
(i'd drop the last sentence as, to me,  problem/solution is specific to [zeta-h123](https://github.com/AxiomMath/zeta-h123) -- i'm not aware of a convention to use this terminology for other lean projects, maybe there is one?)

File: docs/USAGE.md [hunk 2]
```diff
@@ -336,13 +338,20 @@ Example config (`.verilib/probes/config.json`):
 
 You're almost certainly compiling Mathlib from source. probe-lean auto-downloads the cache on first run (see [Mathlib cache](#mathlib-cache-auto-downloaded)), but if that failed, run `lake exe cache get` in the target project manually.
 
-### "environment already contains 'main'"
+### "Co-importability check failed"
+
+Two or more built modules declare the same fully-qualified name — commonly parallel `problem.lean`/`solution.lean` (or spec/impl) files that restate the same definitions without namespaces, or sibling modules that independently define common root-namespace names (`F`, `main`, `digit`, …). The project builds because Lake compiles each module independently, but probe-lean must import **all** built modules into a single Lean environment, and Lean forbids duplicate declarations in one environment (see the co-importability requirement under [Supported Projects](../README.md#supported-projects) in the README).
```
Comment (astefano) [inline docs/USAGE.md]:
```suggestion
Projects with modules that declare the same fully-qualified name build because Lake compiles each module independently, but `probe-lean` must import **all** built modules into a single Lean environment, and Lean forbids duplicate declarations in one environment (see the co-importability requirement under [Supported Projects](../README.md#supported-projects) in the README).
```

File: README.md [hunk 3]
```diff
@@ -34,6 +35,7 @@ Some Lean projects depend on system-level C/C++ libraries (Vulkan, CUDA, OpenSSL
...
 - **Toolchain mismatches** — even a patch-level difference (e.g., v4.28.0-rc1 vs v4.29.0) requires a matching probe-lean build. Use the installer's `--from-project` flag to auto-install the right version
+- **Projects with duplicate declarations across modules** — parallel `problem.lean`/`solution.lean` (or spec/impl) files that restate the same definitions without namespaces, or sibling modules that independently declare common root-namespace names (`F`, `main`, `digit`, …). `lake build` succeeds because Lake compiles modules independently, but the modules cannot coexist in one Lean environment; extraction aborts with a preflight report of the duplicated names (failure signature without the preflight: `environment already contains '<Name>' from <Module>`). Lakefile-level grouping (`defaultTargets`, `[[lean_lib]]` splits) does **not** avoid it — every built `.olean` on disk is analyzed. Fix the project structure (per-family namespaces, or `import` the shared module instead of restating it), or for manual runs extract a non-conflicting subset with `--module <module-name>` (selects the named module plus its submodules)
```
Comment (astefano) [inline README.md]: here we can give the example [zeta-h123](https://github.com/AxiomMath/zeta-h123) to be concrete 
(i'm not sure though what we mean with spec/impl, do we have an example of a lean repo? if not, i'd propose to remove the parenthesis)

File: docs/USAGE.md (hunk 1 above)
Comment (jinxinglim) [reply docs/USAGE.md]: Applied in f83167c — dropped the trailing sentence as suggested.

File: docs/USAGE.md (hunk 2 above)
Comment (jinxinglim) [reply docs/USAGE.md]: Applied verbatim in f83167c.

File: README.md (hunk 3 above)
Comment (jinxinglim) [reply README.md]: Done in f83167c — cited zeta-h123 as the concrete example, dropped the spec/impl parenthesis (no real repo to point to), and restated the general rule as "two built modules declaring the same fully-qualified name" rather than root-namespace-only, since duplicates can live under any namespace.

File: ProbeLean/Coimport.lean [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    full imported constant map (the type's head may live in a dependency),
+    which the preflight doesn't have. We are deliberately lenient there:
+    leniency only under-detects, and a missed collision still fails at import
+    time and hits the fallback hint — whereas strictness could falsely abort
+    a project the importer accepts. -/
+def constSubsumes (a b : ConstantInfo) : Bool :=
+  a.name == b.name &&
+    a.type == b.type &&
+    a.levelParams == b.levelParams &&
+    match a, b with
```
[the remark this answers is not in the document]
Comment (jinxinglim) [reply ProbeLean/Coimport.lean:48]: Keeping the guard — it mirrors the importer rather than contradicting it. In the Lean 4.28 importer (`Lean/Environment.lean`), `finalizeImport` keys its constant map by the declared `cname` from `constNames`, and `subsumesInfo` then *additionally* requires `cinfo₁.name == cinfo₂.name` before tolerating the duplicate. So for the scenario described (same declared name, differing `ConstantInfo.name`s), the importer refuses to subsume and throws — a preflight without the guard would be *more* lenient than the importer and miss that failure. The guard is the faithful choice for the name check specifically (the replica is deliberately lenient elsewhere: the axiom/axiom `isPropCheap` condition needs the full constant map, which the preflight doesn't have — leniency there only under-detects, and misses land on the retained import-failure fallback). Locked with a regression test in f83167c: equal declared name + differing info names + otherwise-identical axioms must collide.

File: ProbeLean/Coimport.lean [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  let mut moduleDecls : Array (Name × Array (Name × ConstantInfo)) := #[]
+  let mut skipped : Array ProjectModule := #[]
+  for m in modules do
+    try
+      -- The CompactedRegion backing the ModuleData is deliberately not freed:
+      -- the ConstantInfo values point into it, and extract is short-lived.
+      -- Only the project's own (small) modules are read here — dependency
+      -- oleans, which dominate memory, are never touched by the preflight.
+      let (data, _) ← readModuleData m.oleanPath
+      moduleDecls := moduleDecls.push (m.name, data.constNames.zip data.constants)
```
[the remark this answers is not in the document]
Comment (jinxinglim) [reply ProbeLean/Coimport.lean:107]: Agreed this is a real scalability consideration — deferring rather than dismissing. To be precise about the tradeoff: the retained regions do compound with the import on the (common) no-collision path; the bound is that the preflight reads only the project's *own* modules and never dependency oleans (Mathlib etc.), which dominate memory in practice, and extract is a short-lived batch process. If a project with heavy own-module data surfaces, the known fix is a two-pass design — collect candidate duplicate names first, then re-read/copy only the fields subsumption needs so each region can be released — noted in the code comment's reasoning. Keeping the simple version until there's evidence it matters.

**Review by astefano — APPROVED** [review]
